CLUE-133 add some documentation on the formula system (#1955)
concord-consortium/codap: v3/doc/formulas.md

#### Example: AttributeFormulaAdapter

- The `AttributeFormulaAdapter` manages all attribute formulas in all datasets.
- Its `getActiveFormulas()` method iterates over all datasets and their attributes, collecting all formulas.
- When recalculating a formula, it uses the formula context and extra metadata to identify and update the correct attribute values in the dataset.

## Formula Dependency Management and Recomputation

Formulas in CODAP can depend on attributes in the same dataset, boundaries, global values, or lookup an attribute from another dataset. The system must track these dependencies to ensure formulas are recomputed when their dependencies change. Also if a dependency is renamed the formula is updated with the new name.

- Each formula's dependencies are determined by parsing its expression and identifying references to dependencies. This is done by `getFormulaDependencies`.
- The Formula Manager watches for changes to the active formulas and then gets the dependencies for any updated formula.

> - Observers are added to the dependencies to trigger recalculation of the formula

Reviewer: ```suggestion
- Observers are added to the dependencies to trigger recalculation of the formula.
```

### Cycles

> Attribute formulas can have dependency cycles, so the FormulaAdapters have a method `getFormulaError` which is used by the Attribute formula adapter to identify this loops and show an error.

Reviewer: ```suggestion
Attribute formulas can have dependency cycles, so the FormulaAdapters have a method `getFormulaError` which is used by the Attribute formula adapter to identify such cycles and show an error.
```
